### PR TITLE
Fix SQL filter generation (prompt engineering), LanguageModelQueryRouter and tests, update dependencies and tests

### LINKED ISSUE #2253: [BUG] LanguageModelSqlFilterBuilderIT does not support EXTRACT(...) expression
**Describe the bug**

LanguageModelSqlFilterBuilderIT test is failing due unsupported predicate expression: `EXTRACT(year FROM NOW())`

Apparently, that [LLM prompt](https://github.com/langchain4j/langchain4j/blob/1acb7a607f9eeaa9a5caf04284bd96f233004670/embedding-store-filter-parsers/langchain4j-embedding-store-filter-parser-sql/src/main/java/dev/langchain4j/store/embedding/filter/builder/sql/LanguageModelSqlFilterBuilder.java#L90-L105) should be more specific about supported SQL predicates

**Log and Stack trace**

OpenAI has generated SQL filter that is not supported by the LanguageModelSqlFilter:
```sql
SELECT name, genre, year
FROM movies
WHERE year = EXTRACT(year FROM NOW()) - 1;
```
Failed build: https://github.com/langchain4j/langchain4j/actions/runs/12236022874/attempts/1

**To Reproduce**
Can be reproduced with SqlFilterParserTest, if `EXTRACT(...)` should be supported or with LanguageModelSqlFilterBuilderIT, if prompt should be updated.

**Expected behavior**
LLM generates only supported predicates.

**Please complete the following information:**
- LangChain4j version: 0.37.0-SNAPSHOT
- LLM(s) used: OpenAI
- Java version: 21, 17
- Spring Boot version (if applicable): -

**Additional context**

The test might be flaky test due to undetermined LLM behaviour

### PR DESCRIPTION
**Pull request by kpavlov**
## Issue

Fixes #2253 

## Change

- Improved SQL query generation rules in the `LanguageModelSqlFilterBuilder` by adding injection prevention and ANSI SQL compliance. Updated `jsqlparser` to version 4.9 and removed an unused test dependency (`mockito-core`) for cleanup. These changes enhance security, maintainability, and adherence to standards.
- Replaced flaky `AiServicesWithRagIT.should_fail_when_query_is_ambiguous`  with unit test.
- Updated LanguageModelQueryRouter and tests

### Dependency Updates:
* Updated `jsqlparser` version from 4.8 to 4.9 in `pom.xml`.
* Changed `junit-jupiter-engine` to `junit-jupiter-api` and added `mockito-junit-jupiter` dependency for testing in `pom.xml`. [[1]](diffhunk://#diff-568a04ebfbcb6d90a38652244128471294f4a0a26bb153b79f1a628c74d71d0dL37-R37) [[2]](diffhunk://#diff-568a04ebfbcb6d90a38652244128471294f4a0a26bb153b79f1a628c74d71d0dR53-R58)
* Removed the `mockito-core` dependency in `pom.xml`.

### SQL Query Generation Improvements:
* Enhanced the default prompt template in `LanguageModelSqlFilterBuilder` to prevent SQL injection, use ANSI SQL functions, and avoid line breaks.
* Simplified the `build` method to directly trim the generated SQL and log both the cleaned and original SQL on parsing failure.

### Unit Test Enhancements:
* Added `@EnabledIfEnvironmentVariable` annotation and improved assertions in `LanguageModelSqlFilterBuilderIT`. [[1]](diffhunk://#diff-6e40d8bdab35b72e7241bf75062d06980383348382aa0baa2539de7d47f9ceb3R3-R18) [[2]](diffhunk://#diff-6e40d8bdab35b72e7241bf75062d06980383348382aa0baa2539de7d47f9ceb3L29-R30) [[3]](diffhunk://#diff-6e40d8bdab35b72e7241bf75062d06980383348382aa0baa2539de7d47f9ceb3L65-R68) [[4]](diffhunk://#diff-6e40d8bdab35b72e7241bf75062d06980383348382aa0baa2539de7d47f9ceb3L81-R91) [[5]](diffhunk://#diff-6e40d8bdab35b72e7241bf75062d06980383348382aa0baa2539de7d47f9ceb3L100-R119)
* Updated `LanguageModelSqlFilterBuilderTest` to use Mockito for mocking dependencies and verifying interactions. [[1]](diffhunk://#diff-95d3dc532dfe6c96f1baa66b19e238538f77f141c79a8c1d872fdce4ce759c43R3-R32) [[2]](diffhunk://#diff-95d3dc532dfe6c96f1baa66b19e238538f77f141c79a8c1d872fdce4ce759c43R53-R64) [[3]](diffhunk://#diff-95d3dc532dfe6c96f1baa66b19e238538f77f141c79a8c1d872fdce4ce759c43R95-R102)

## General checklist

- [ ] There are no breaking changes
- [ ] I have added unit and/or integration tests for my change
- [ ] The tests cover both positive and negative cases
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by kpavlov** [comment]
@langchain4j , integration tests are failing due to "Your credit balance is too low to access the Anthropic API. Please go to Plans & Billing to upgrade or purchase credits."

**Comment by langchain4j** [comment]
Hey @kpavlov, is it relevant to this PR? Does it block you?

**Comment by kpavlov** [comment]
Integration tests in all other PRs are failing due to [#2253](https://github.com/langchain4j/langchain4j/issues/2253), that's why I think it is unblocker. The failing Anthropic tests are not relevant to this PR indeed. Only you with a Super Power can merge it:)

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-parser-sql/src/main/java/dev/langchain4j/store/embedding/filter/builder/sql/LanguageModelSqlFilterBuilder.java [hunk 1]
```diff
@@ -136,13 +138,13 @@ public Filter build(Query query) {
 
         String generatedSql = response.content().text();
 
-        String cleanedSql = clean(generatedSql);
+        String cleanedSql = generatedSql.trim();
```
Comment (langchain4j) [inline embedding-store-filter-parsers/langchain4j-embedding-store-filter-parser-sql/src/main/java/dev/langchain4j/store/embedding/filter/builder/sql/LanguageModelSqlFilterBuilder.java:141]: This is potentially breaking change: `clean(String)` is there to let users to extend this class and override the cleaning behavior (if required)

File: langchain4j-core/src/main/java/dev/langchain4j/rag/query/router/LanguageModelQueryRouter.java [hunk 2]
```diff
@@ -103,7 +102,14 @@ public Collection<ContentRetriever> route(Query query) {
         Prompt prompt = createPrompt(query);
         try {
             String response = chatLanguageModel.generate(prompt.text());
-            return parse(response);
+
+            final var routes = parse(response);
+            if (routes.isEmpty()) {
+                log.warn("No routes found for query: '{}'", query.text());
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/rag/query/router/LanguageModelQueryRouter.java]:
```suggestion
                log.warn("No content retrievers found for query: '{}'", query.text());
```

File: langchain4j-core/src/main/java/dev/langchain4j/rag/query/router/LanguageModelQueryRouter.java [hunk 3]
```diff
@@ -120,7 +126,7 @@ protected Collection<ContentRetriever> fallback(Query query, Exception e) {
                 log.debug("Fallback: query '{}' will be routed to all available content retrievers", query.text());
                 yield new ArrayList<>(idToRetriever.values());
             }
-            default -> throw new RuntimeException(e);
+            case FAIL -> throw new RuntimeException("Failed to route query", e);
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/rag/query/router/LanguageModelQueryRouter.java:129]: nit: perhaps add a query text?

**Review by ghost — COMMENTED** [review]
@kpavlov thanks a lot!

**Comment by kpavlov** [comment]
Let's start with #2294, since it's smaller and there is less chance to break something

File: embedding-store-filter-parsers/langchain4j-embedding-store-filter-parser-sql/src/main/java/dev/langchain4j/store/embedding/filter/builder/sql/LanguageModelSqlFilterBuilder.java (hunk 1 above)
Comment (kpavlov) [reply embedding-store-filter-parsers/langchain4j-embedding-store-filter-parser-sql/src/main/java/dev/langchain4j/store/embedding/filter/builder/sql/LanguageModelSqlFilterBuilder.java:141]: Let's drop this and take #2294 then

**Comment by kpavlov** [comment]
Let's use this one instead #2311

File: langchain4j-core/src/main/java/dev/langchain4j/rag/query/router/LanguageModelQueryRouter.java (hunk 3 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/rag/query/router/LanguageModelQueryRouter.java:129]: Done in #2311

**Pull request closed without merging**